doc/adr/0008-jetstream-kv.md

```diff
@@ -0,0 +1,326 @@
+# 8. JetStream based Key-Value Stores
+
+|Metadata|Value|
+|--------|-----|
+|Date    |2021-06-30|
+|Author  |@ripienaar|
+|Status  |Partially Implemented|
+|Tags    |jetstream, client, kv|
+
+## Context
+
+This document describes a design and initial implementation of a JetStream backed key-value store. The initial implementation
+is available in the CLI as `nats kv` with an initial reference client package implementation in `jsm.go` repository to support
+the CLI.
+
+This document aims to guide client developers in implementing this feature in the language clients we maintain.
+
+## Overview
+
+We intend to hit a basic initial feature set as below, with some future facing goals as indicated:
+
+Initial feature list:
+
+ * Multiple named buckets full of keys with n historical values kept per value
+ * Put and Get of `string(k)=string(v)` values
+ * Put only if the sequence of the last value for a key matches an expected sequence
+ * Historic values can be kept per key, setable on creation of the bucket
+ * Key deletes preserves history
```

Comment:
Instead of deleting all messages for KV.BUCKET.x we put a message in saying its a delete operation.  Historical values are kept, client will return key not found on Get() if the last value for that subject is a delete op.

Reply:
Why would you want that? Just curious.

Reply:
A if I tell you I keep 5 historic values, then I should keep 5 historic values when if its 4 old + delete.  I think its important to be able to undo a delete.

On a more practical level, if I delete a subject with our new capabilities anyone consuming that stream - perhaps on a filter subjet - has no idea this happened.  This present a problem for things like the cache, because it will never evict a deleted key from the cache.

So this solves the cache consistency issue - but more importantly I think it preserved the idea behind historical values.